GVRf/Framework/jni/shaders/material/lightmap_shader.cpp

```diff
@@ -0,0 +1,118 @@
+        "attribute vec3 a_normal;\n"
+        "attribute vec2 a_tex_coord;\n"
+        "uniform mat4 u_mvp;\n"
+        "varying vec2 coord;\n"
+        "void main() {\n"
+        " vec4 pos = u_mvp * a_position;\n"
+        " coord = a_tex_coord;\n"
+        " gl_Position = pos;\n"
+        "}";
+
+static const char FRAGMENT_SHADER[] = "precision mediump float;\n"
+        "varying vec2  coord;\n"
+        "uniform sampler2D u_main_texture;\n"
+        "uniform sampler2D u_lightmap_texture;\n"
+        "uniform vec2  u_lightmap_offset;\n"
+        "uniform vec2  u_lightmap_scale;\n"
+        "void main() {\n"
+        " vec4 color;\n"
+        " vec4 lightmap_color;\n"
+        " vec2 lightmap_coord = (coord * u_lightmap_scale) + u_lightmap_offset;\n"
+        // Beast exports the texture with vertical flip
+        " lightmap_color = texture2D(u_lightmap_texture, vec2(lightmap_coord.x, 1.0 - lightmap_coord.y));\n"
+        " color = texture2D(u_main_texture, coord);\n"
+        " gl_FragColor = color * lightmap_color;\n"
+        "}";
+
+LightMapShader::LightMapShader() :
+        program_(0), u_mvp_(0), u_texture_(0), u_lightmap_texture_(0),
+        u_lightmap_offset_(0), u_lightmap_scale_(0) {
+    program_ = new GLProgram(VERTEX_SHADER, FRAGMENT_SHADER);
+    u_mvp_ = glGetUniformLocation(program_->id(), "u_mvp");
+    u_texture_ = glGetUniformLocation(program_->id(), "u_texture");
+
+    u_lightmap_texture_ = glGetUniformLocation(program_->id(), "u_lightmap_texture");
+    u_lightmap_offset_ = glGetUniformLocation(program_->id(), "u_lightmap_offset");
+    u_lightmap_scale_ = glGetUniformLocation(program_->id(), "u_lightmap_scale");
+}
+
+LightMapShader::~LightMapShader() {
+    delete program_;
+}
+
+void LightMapShader::render(const glm::mat4& mvp_matrix,
+        RenderData* render_data, Material* material) {
+    Mesh* mesh = render_data->mesh();
+    Texture* texture = material->getTexture("main_texture");
+    Texture* lightmap_texture = material->getTexture("lightmap_texture");
+    glm::vec2 lightmap_offset = material->getVec2("lightmap_offset");
+    glm::vec2 lightmap_scale = material->getVec2("lightmap_scale");
+
+#if _GVRF_USE_GLES3_
+    mesh->generateVAO();
+#endif
+
+    glUseProgram(program_->id());
+
+#if not _GVRF_USE_GLES3_
+    glVertexAttribPointer(a_position_, 3, GL_FLOAT, GL_FALSE, 0,
+            mesh->vertices().data());
+    glEnableVertexAttribArray(a_position_);
+#endif
+
+    glUniformMatrix4fv(u_mvp_, 1, GL_FALSE, glm::value_ptr(mvp_matrix));
+
+    glActiveTexture (GL_TEXTURE0);
+    glBindTexture(texture->getTarget(), texture->getId());
+    glUniform1i(u_texture_, 0);
+
+    glActiveTexture (GL_TEXTURE1);
+    glBindTexture(lightmap_texture->getTarget(), lightmap_texture->getId());
+    glUniform1i(u_lightmap_texture_, 1);
+
+    glUniform2f(u_lightmap_offset_, lightmap_offset.x, lightmap_offset.y);
+    glUniform2f(u_lightmap_scale_, lightmap_scale.x, lightmap_scale.y);
+
+#if _GVRF_USE_GLES3_
```

Comment:
I think it's OK. We haven't tested GLES 2 for a long time. I don't think it is required any more. @thomasflynn Do you have more information about it?